lib/vec_deque.c

```diff
@@ -0,0 +1,101 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "vec_deque.h"
+
+typedef struct VecDeque {
+  size_t capacity;
+  void** array;
+  size_t start;
+  size_t end;
+  size_t len;
+} Node;
+
+static void resize_if_needed(VecDeque* deque) {
+  if (deque->capacity == deque->len) {
+    vec_deque_reserve(deque, deque->capacity);
+  }
+}
+
+static void rot_left(size_t* v, size_t cap) {
+  *v  = (cap + *v - 1) % cap;
+}
+
+static void rot_right(size_t* v, size_t cap) {
+  *v = (*v + 1) % cap;
+}
+
+VecDeque* vec_deque_init(size_t capacity) {
+  if (capacity == 0) {
```

Comment:
optional feature request: could we also support initial-capacity of zero?

Reply:
Done! :D